internal/app/referenceclient/wire_details.go

```diff
@@ -0,0 +1,321 @@
+		// A nil response in the trace is valid if the HTTP round trip failed.
+		return 0, false
+	}
+
+	// Check end-stream and/or error JSON data in the response.
+	contentType := trace.Response.Header.Get("content-type")
+	switch {
+	case isUnaryJSONError(contentType, statusCode):
+		// If this is a unary request that returned an error, then use the entire
+		// response body as the wire error details.
+		examineConnectError(wrapper.buf.Bytes(), printer)
+	case strings.HasPrefix(contentType, "application/connect+"):
+		// If this is a streaming Connect request, then look through the trace events
+		// for the ResponseBodyEndStream event and parse its content into an
+		// endStreamError to see if there are any error details.
+		endStreamContent, ok := getBodyEndStream(trace)
+		if ok {
+			examineConnectEndStream([]byte(endStreamContent), printer)
+		}
+	case strings.HasPrefix(contentType, "application/grpc-web"):
+		// For gRPC-Web, capture the trailers in the body. We don't do any case normalization
+		// or trimming of excess whitespace so that the full values are available to check.
+		endStreamContent, ok := getBodyEndStream(trace)
+		if ok {
+			examineGRPCEndStream(endStreamContent, printer)
+		}
+	}
+
+	if contentType != "application/grpc" && !strings.HasPrefix(contentType, "application/grpc+") {
+		// It's not gRPC protocol, so there should be no HTTP trailers.
+		if len(trace.Response.Trailer) > 0 {
+			printer.Printf("response included %d HTTP trailers but should not have any", len(trace.Response.Trailer))
+		}
+	}
+
+	return trace.Response.StatusCode, true
+}
+
+type wireTracer struct {
+	tracer *tracer.Tracer
+}
+
+// Complete intercepts the Complete call for a tracer, extracting wire details
+// from the passed trace. The wire details will be stored in the context acquired by
+// withWireCapture and can be retrieved via examineWireDetails.
+func (t *wireTracer) Complete(trace tracer.Trace) {
+	setWireTrace(trace.Request.Context(), &trace)
+
+	if t.tracer != nil {
+		t.tracer.Complete(trace)
+	}
+}
+
+// isUnaryJSONError returns whether the given content type and HTTP status code
+// represents a unary JSON error.
+func isUnaryJSONError(contentType string, statusCode int) bool {
+	return contentType == "application/json" && statusCode != 200
+}
+
+// getBodyEndStream returns the contents of any end-stream message in the trace.
+// The bool value will be true if an end-stream message is found and otherwise
+// it will be false.
+func getBodyEndStream(trace *tracer.Trace) (string, bool) {
+	for _, event := range trace.Events {
+		endStream, ok := event.(*tracer.ResponseBodyEndStream)
+		if !ok {
+			continue
+		}
+		return endStream.Content, true
+	}
+	return "", false
+}
+
+func examineConnectError(_ json.RawMessage, _ internal.Printer) {
+	// TODO
+}
+
+func examineConnectEndStream(_ json.RawMessage, _ internal.Printer) {
+	// TODO
+}
+
+func examineGRPCEndStream(endStream string, printer internal.Printer) {
```

Comment:
Holy cow, haha, this is great.  This almost feels like it could be its own package to help others validate end streams.